python-package/xgboost/utils.py

```diff
@@ -0,0 +1,162 @@
+# pylint: disable=c-extension-no-member
+"""Various helper functions."""
+
+import math
+import os
+import warnings
+from typing import TYPE_CHECKING, List, Optional, Sequence, Tuple
+
+import numpy as np
+
+if TYPE_CHECKING:
+    from cuda.bindings import runtime as cudart
+
+_MASK_SIZE = 64
+
+
+class _BitField64:
+    """A simplified version of the bit field in XGBoost C++."""
+
+    def __init__(self, mask: Sequence) -> None:
+        self.mask: List[int] = []
+        for m in mask:
+            self.mask.append(m)
+
+    @staticmethod
+    def to_bit(i: int) -> Tuple[int, int]:
+        """Split the index into bit position and value position."""
+        int_pos, bit_pos = 0, 0
+        if i == 0:
+            return int_pos, bit_pos
+
+        int_pos = i // _MASK_SIZE
+        bit_pos = i % _MASK_SIZE
+        return int_pos, bit_pos
+
+    def check(self, i: int) -> bool:
+        """Check whether the i bit is set."""
+        ip, bp = self.to_bit(i)
+        value = self.mask[ip]
+        test_bit = 1 << bp
+        res = value & test_bit
+        return bool(res)
+
+
+def _checkcu(status: "cudart.cudaError_t") -> None:
+    from cuda.bindings import runtime as cudart
+
+    if status != cudart.cudaError_t.cudaSuccess:
+        raise RuntimeError(cudart.cudaGetErrorString(status))
+
+
+def _get_uuid(ordinal: int) -> str:
+    """Construct a string representation of UUID."""
+    from cuda.bindings import runtime as cudart
+
+    status, prop = cudart.cudaGetDeviceProperties(ordinal)
+    _checkcu(status)
+
+    dash_pos = {0, 4, 6, 8, 10}
+    uuid = "GPU"
+
+    for i in range(16):
+        if i in dash_pos:
+            uuid += "-"
+        h = hex(0xFF & np.int32(prop.uuid.bytes[i]))
+        assert h[:2] == "0x"
+        h = h[2:]
+
+        while len(h) < 2:
+            h = "0" + h
+        uuid += h
+    return uuid
+
+
+def _get_ordinal(device: Optional[str]) -> int:
+    if device is None:
+        device = "cuda"
+
+    split = device.split(":")
+    if len(split) == 1:
+        ordinal = 0
+    elif len(split) == 2:
+        ordinal = int(split[1])
+    else:
+        raise ValueError(f"Invalid device: {device}")
+    return ordinal
```

Comment:
Seems overly complicated, why not just take an integer representing an index or the UUID? Also done in https://github.com/rapidsai/dask-cuda/blob/6c3223d7f82f4b2e7eb209a8485b1e32cb5357a7/dask_cuda/utils.py#L752-L775.

Reply:
Primarily for consistency with XGBoost's notation for devices: "cuda:0", "cpu", I'm trying not to come up with a new one.

Reply:
Ah ok, makes sense. I didn't remember how XGBoost notation looked like.